clang/test/CheckedC/static-checking/bounds-decl-checking.c

```diff
@@ -286,7 +288,37 @@ int f30(_Ptr<struct S> p) {
   return 0;
 }
 
-int f31(_Ptr<void> p) {
+int f31(_Array_ptr<struct S> p : count(1), _Array_ptr<struct R> q : count(1)) { // expected-note {{(expanded) declared bounds are 'bounds(p, p + 1)'}}
+  // We cannot compare unequal incomplete referent types 'struct S' and 'struct R' of p and q.
+  p = (_Array_ptr<struct S>)q; // expected-warning {{cannot prove declared bounds for 'p' are valid after assignment}} \
+                               // expected-note {{(expanded) inferred bounds are 'bounds(q, q + 1)'}}
+}
+
+int f32(_Array_ptr<struct S> p : count(0), _Array_ptr<struct S> q : count(1)) {
+  // p and q have the same incomplete referent type 'struct S'.
+  p = q;
+}
+
+int f33(_Ptr<struct S> p) {
+  test_f31(p);
+  return 0;
+}
+
+int f34(_Array_ptr<struct S> p : bounds(p, p), _Array_ptr<struct S> q : count(0)) { // expected-note {{(expanded) declared bounds are 'bounds(p, p)'}}
+  // The upper bound q + 0 is an incomplete constant sized upper offset,
+  // but the upper bound p is not an incomplete constant sized upper offset.
+  p = q; // expected-warning {{cannot prove declared bounds for 'p' are valid after assignment}} \
```

Comment:
I think the expected warning on line 310 in this test case is not really correct. The lower/upper bounds of `p` are not marked as `Lower/UpperConstantIncomplete` because of the way these fields are set. Currently, these fields are set only when the call to `GetReferentSizeInChars` fails. In the case of the bounds expression `(p, p)`, I guess the control is going to the `exit` label in `SplitIntoBaseAndOffset` where the lower and upper bounds are split as `p + 0` without calling `GetReferentSizeInChars`, and therefore the incompleteness of the type of `p` is not recorded.

I suggest the following:
Try checking if the type of `Base` is incomplete in the code under the `exit` label in `SplitIntoBaseAndOffset`. If the type is incomplete, return `Kind::IncompleteConstantSized`.  If this gets further complicated or causes more issues/failures, file an issue for this test case and adding a TODO here with a reference to the filed issue. 

Ideally, a `BaseRange` object must be marked as having an incomplete type if the type of the base (in this case `p`) is incomplete.